code/game/turfs/simulated/floor_icon.dm

```diff
@@ -132,28 +140,171 @@ var/list/flooring_cache = list()
+
+		//If is_hole is true, then it's space or openspace
+		else if(T.is_hole)
+			if(space_smooth)
+				is_linked = TRUE
+
+
+		//If we get here then its a normal floor
+		else if (istype(T, /turf/simulated/floor))
+			var/turf/simulated/floor/t = T
+			//If the floor is the same as us,then we're linked,
+			if (t.flooring.name == name)
+				/*
+					But there's a caveat. To make atom black/whitelists work correctly, we also need to check that
+					they smooth with us. Ill call this counterchecking for simplicity.
+					This is needed to make both turfs have the correct borders
+
+					To prevent infinite loops we have a countercheck var, which we'll set true
+				*/
+				if (smooth_movable_atom != SMOOTH_NONE)
+					//We do the countercheck, passing countercheck as true
+					is_linked = test_link(T, origin, countercheck = TRUE)
+
+
+			else if (floor_smooth == SMOOTH_ALL)
+				is_linked = TRUE
+
+			else if (floor_smooth != SMOOTH_NONE)
+
+
+
+				//If we get here it must be using a whitelist or blacklist
+				if (floor_smooth == SMOOTH_WHITELIST)
+					for (var/v in flooring_whitelist)
+						if (istype(t.flooring, v))
+							//Found a match on the list
+							is_linked = TRUE
+							break
+				else if(floor_smooth == SMOOTH_BLACKLIST)
+					is_linked = TRUE //Default to true for the blacklist, then make it false if a match comes up
+					for (var/v in flooring_whitelist)
+						if (istype(t.flooring, v))
+							//Found a match on the list
+							is_linked = FALSE
+							break
+	else
+		//If this is a countercheck, we skip all of the above, start off with true, and go straight to the atom lists
+		is_linked = TRUE
+
+
+
+	//Alright now we have a preliminary answer about smoothing, however that answer may change with the following
+	//Atom lists!
+	var/best_priority = -1
+	//A white or blacklist entry will only override smoothing if its priority is higher than this
+	//And then this value becomes its priority
+	if (smooth_movable_atom != SMOOTH_NONE)
+		if (smooth_movable_atom == SMOOTH_WHITELIST || smooth_movable_atom == SMOOTH_GREYLIST)
+			for (var/list/v in movable_atom_whitelist)
+				var/d_type = v[1]
+				var/list/d_vars = v[2]
+				var/d_priority = v[3]
+				//Priority is the quickest thing to check first
+				if (d_priority <= best_priority)
+					continue
+
+				//Ok, now we start testing all the atoms in the target turf
+				for (var/a in T) //No implicit typecasting here, faster
+
+					if (istype(a, d_type))
+						//It's the right type, so we're sure it will have the vars we want.
+
+						var/atom/movable/AM = a
+						//Typecast it to a movable atom
+						//Lets make sure its in the way before we consider it
+						if (!AM.is_between_turfs(origin, T))
+							continue
+
+						//From here on out, we do dangerous stuff that may runtime if the coder screwed up
+
+
+						var/match = TRUE
+						for (var/d_var in d_vars)
+							//For each variable we want to check
+							if (AM.vars[d_var] != d_vars[d_var])
```

Comment:
`AM.vars[d_var]` will runtime if `d_var` not in `AM.vars`

Reply:
I'm aware, the comment indicates as such.
But it will be in there, because the type is already checked beforehand.

Reply:
That's only part of problem.
That specified type can be change but changer will have no clue that it vars uses here.